### PR TITLE
Add fetch_identity_path logic for identity from token flow

### PR DESCRIPTION
**Pull request by nessiLahav**
### What does this PR do?
Add new identity-path jwt authenticator configuration 

The identity-path is relevant only for token identity flow and doesn't change the user in url flow.
If identity-path not configured the default value is empty which means hosts in root policy.

Note: This feature supports only hosts and blocks users

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

### CODE REVIEW

**Comment by tzheleznyak** [comment]
Add JIRA issue to commit and mark checkboxes

File: app/domain/authentication/authn_jwt/consts.rb [hunk 1]
```diff
@@ -11,6 +11,10 @@ module AuthnJwt
     IDENTITY_NOT_RETRIEVED_YET = "Identity not retrieved yet"
     URL_IDENTITY_PROVIDER_INTERFACE_NAME = "url-identity-provider"
     TOKEN_IDENTITY_PROVIDER_INTERFACE_NAME = "token-identity-provider"
+    IDENTITY_PATH_RESOURCE_NAME = "identity-path"
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/consts.rb:14]: @hilagross what should be the variable name?

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
+require 'command_class'
+
+module Authentication
+  module AuthnJwt
+    module IdentityProviders
+      # Add policy prefix to identity, for examples:
+      # 1. Input: identity_prefix=host identity=host_test1
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:7]: I think that if we have UT coverage the examples are redundant here

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+          if identity_prefix_last_character != IDENTITY_PATH_CHARACTER_DELIMITER
+            @identity_prefix_with_character_suffix = @identity_prefix + IDENTITY_PATH_CHARACTER_DELIMITER
+          else
+            @identity_prefix_with_character_suffix = @identity_prefix
+          end
+        end
+
+        def identity_prefix_last_character
+          @identity_prefix_last_character ||= @identity_prefix[-1, 1]
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:69]:
```suggestion
          @identity_prefix_last_character ||= @identity_prefix[-1]
```
Can you try this?

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      # 4. Input: identity_prefix=host identity=/host_test1
+      #    Output: host/host_test1
+      AddPrefixToIdentity = CommandClass.new(
+        dependencies: {
+          logger: Rails.logger
+        },
+        inputs: %i[identity_prefix identity]
+      ) do
+        def call
+          validate_identity_exists
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:22]: Move this two validate call to a function called validate input that calls both of them

File: app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      ) do
+        def call
+          fetch_identity_path
+        end
+
+        private
+
+        def fetch_identity_path
+          @logger.debug(LogMessages::Authentication::AuthnJwt::FetchingIdentityPath.new)
+          validate_configuration_and_set_identity_path_value
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb:23]: Move validate and logs to call functioion

File: app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        def set_identity_path_from_policy
+          @identity_path = identity_path_required_secret_value
+        end
+
+        def identity_path_required_secret_value
+          @identity_path_required_secret_value ||= identity_path_required_secret[identity_path_resource_id]
+        end
+
+        def identity_path_required_secret
+          @identity_path_required_secret ||= @fetch_required_secrets.(resource_ids: [identity_path_resource_id])
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb:69]: Lets use call for command classes in all our authenticator. Currently only in jwt and later we will change other authenticators in R&D boost

File: app/domain/errors.rb [hunk 7]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
         code: "CONJ00098E"
       )
 
+      MissingIdentity = ::Util::TrackableErrorClass.new(
```
Comment (tzheleznyak) [inline app/domain/errors.rb:477]: @shulifink May you please review this error messages?

CC: @hilagross

File: app/domain/logs.rb [hunk 8]
```diff
@@ -555,6 +555,51 @@ module AuthnJwt
         code: "CONJ00109D"
       )
 
+      FetchingIdentityPath = ::Util::TrackableLogMessageClass.new(
```
Comment (nessiLahav) [inline app/domain/logs.rb:558]: @shulifink can you review the following logs ?

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb (hunk 2 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:7]: I think it is more easy to understand the logic for the comment

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb (hunk 3 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:69]: OK

File: app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb (hunk 5 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/fetch_identity_path.rb:23]: as we agreed i ignore this

File: app/domain/authentication/authn_jwt/consts.rb (hunk 1 above)
Comment (Avi-Eyal) [reply app/domain/authentication/authn_jwt/consts.rb:14]: @tzheleznyak I think that the variable name should stay `identity-path`

Comment (benoaviram) [reply app/domain/authentication/authn_jwt/consts.rb:14]: indeed this is the variable name

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+          return @identity_prefix_with_character_suffix if @identity_prefix_with_character_suffix
+
+          if identity_prefix_last_character != IDENTITY_PATH_CHARACTER_DELIMITER
+            @identity_prefix_with_character_suffix = @identity_prefix + IDENTITY_PATH_CHARACTER_DELIMITER
+          else
+            @identity_prefix_with_character_suffix = @identity_prefix
+          end
+
+          @identity_prefix_with_character_suffix
+        end
```
Comment (micahlee) [inline app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:72]: You can simplify this method (and appease Code Climate) with:

```suggestion
        def identity_prefix_with_delimiter_suffix
          @identity_prefix_with_delimiter_suffix ||=
            if identity_prefix_last_character != IDENTITY_PATH_CHARACTER_DELIMITER
              @identity_prefix + IDENTITY_PATH_CHARACTER_DELIMITER
            else
              @identity_prefix
          end
        end
```

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb [hunk 10]
```diff
@@ -7,69 +7,154 @@ def initialize(
...
+              )
+          end
+
+          @identity_name_with_prefix_path
+        end
+
+        def identity_name_with_prefix_path
+          @identity_name_with_prefix_path ||= add_prefix_path_to_identity_name
+        end
+
```
Comment (micahlee) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:139]: Here is another example that should be simplified and can remove the `add_prefix_path_to_identity_name` method altogether. You may be aware of others following this same pattern that I haven't spotted. These should be cleaned up.

```suggestion
           def identity_name_with_prefix_path
          @identity_name_with_prefix_path ||=
            if identity_path.blank?
              identity_name_from_token
            else
              @add_prefix_to_identity.call(
                identity_prefix: identity_path,
                identity: identity_name_from_token
              )
            end
        end

```

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb [hunk 11]
```diff
@@ -7,69 +7,154 @@ def initialize(
...
+        def add_host_to_identity_with_prefix
+          host_identity_with_prefix
+        end
```
Comment (micahlee) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:142]: Why do you need this method? It looks like you should be able to use `host_identity_with_prefix` directly.

File: app/domain/errors.rb [hunk 12]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
...
+        msg: "Identity is empty or not found.",
```
Comment (shulifink) [inline app/domain/errors.rb:478]: Can we give the identity an identifier? I;m giessing this is the JWT identity??
 ```suggestion
        msg: "JWT identity is empty or was not found.",
```

File: app/domain/errors.rb [hunk 13]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
...
+        code: "CONJ00101E"
+      )
+
+      MissingIdentityPrefix = ::Util::TrackableErrorClass.new(
+        msg: "Identity prefix is empty or not found.",
```
Comment (shulifink) [inline app/domain/errors.rb:483]:
```suggestion
        msg: "JWT identity prefix is empty or was not found.",
```

File: app/domain/logs.rb [hunk 14]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        msg: "Fetching identity path...",
+        code: "CONJ00110D"
+      )
+
+      FetchedIdentityPath = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully fetched identity path '{0-identity-path}'",
```
Comment (shulifink) [inline app/domain/logs.rb:564]:
```suggestion
        msg: "Successfully fetched JWT identity path '{0-identity-path}'",
```

File: app/domain/logs.rb [hunk 15]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00111I"
+      )
+
+      FetchingIdentityByInterface = ::Util::TrackableLogMessageClass.new(
+        msg: "Fetching identity by interface: '{0-interface-name}'",
```
Comment (shulifink) [inline app/domain/logs.rb:569]:
```suggestion
        msg: "Fetching JWT identity by interface: '{0-interface-name}'...",
```
what do you mean by "interface"? Can you provide an example?

File: app/domain/logs.rb [hunk 16]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00112D"
+      )
+
+      FetchedIdentityByInterface = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully fetched identity '{0-identity}' by interface: '{1-interface-name}'",
```
Comment (shulifink) [inline app/domain/logs.rb:574]: Can you please provide example?

File: app/domain/logs.rb [hunk 17]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00113I"
+      )
+
+      AddingIdentityPrefixToIdentity = ::Util::TrackableLogMessageClass.new(
+        msg: "Adding identity prefix '{0-identity-prefix}' to identity '{1-identity}'...",
```
Comment (shulifink) [inline app/domain/logs.rb:579]: Example, please?

File: app/domain/logs.rb [hunk 18]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00114D"
+      )
+
+      AddedIdentityPrefixToIdentity = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully added identity prefix, calculated identity is: '{0-identity}'",
```
Comment (shulifink) [inline app/domain/logs.rb:584]:
```suggestion
        msg: "Successfully added JWT identity prefix. Calculated identity is: '{0-identity}'",
```

File: app/domain/logs.rb [hunk 19]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00115D"
+      )
+
+      IdentityPathNotConfigured = ::Util::TrackableLogMessageClass.new(
+        msg: "Identity path '{0-resource-name}' is not configured, the identity will be taken from root policy",
```
Comment (shulifink) [inline app/domain/logs.rb:589]:
```suggestion
        msg: "JWT identity path '{0-resource-name}' not configured. JWT identity will be taken from root policy",
```

**Review by shulifink — CHANGES_REQUESTED** [review]

File: app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb (hunk 9 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/add_prefix_to_identity.rb:72]: Nice , thanks :)

File: app/domain/authentication/authn_jwt/consts.rb [hunk 20]
```diff
@@ -11,6 +11,10 @@ module AuthnJwt
...
+    IDENTITY_PATH_DEFAULT_VALUE = ""
+    IDENTITY_PATH_CHARACTER_DELIMITER = "/"
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/consts.rb:16]: Is there another const that defines the character for all Conjur globally?

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb (hunk 10 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:139]: Nice , thanks :)

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb (hunk 11 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:142]: It is just for telling the story of this class in the main function `jwt_identity`

```
          fetch_identity_name_from_token
          fetch_identity_path
          add_prefix_path_to_identity_name
          add_host_to_identity_with_prefix
```

File: app/domain/logs.rb (hunk 15 above)
Comment (nessiLahav) [reply app/domain/logs.rb:569]: Technical term which indicate on a specific flow in the code

```
CONJ00058D Selecting identity provider interface...
CONJ00059I Selected identity provider interface: 'token-identity-provider'
CONJ00114D Fetching identity by interface: 'token-identity-provider'
```

We currently have to interfaces in the code: 'token-identity-provider' 'url-identity-provider' which indicating from where we will fetch the identity of the token.

In the future we may have more interface implementations

File: app/domain/logs.rb (hunk 16 above)
Comment (nessiLahav) [reply app/domain/logs.rb:574]:
```
CONJ00115I Successfully fetched identity 'host/apps/00uksxz3o2vFJCX230h7' by interface: 'token-identity-provider'
```

File: app/domain/logs.rb (hunk 17 above)
Comment (nessiLahav) [reply app/domain/logs.rb:579]: identity-path is a variable in the authenticator policy which defined the host identity path(location in the policy tree),
in the following example:
the identity in the token is: 00uksxz3o2vFJCX230h7
the customer configured the identiy-path to hosts in path /apps/
the host definition in Conjur is : 'host/apps/00uksxz3o2vFJCX230h7

```
 CONJ00111I Successfully fetched identity path 'apps'
 CONJ00116D Adding identity prefix 'apps' to identity '00uksxz3o2vFJCX230h7'...
 CONJ00117D Successfully added identity prefix, calculated identity is: 'apps/00uksxz3o2vFJCX230h7'
 CONJ00116D Adding identity prefix 'host' to identity 'apps/00uksxz3o2vFJCX230h7'...
 CONJ00117D Successfully added identity prefix, calculated identity is: 'host/apps/00uksxz3o2vFJCX230h7'
 CONJ00115I Successfully fetched identity 'host/apps/00uksxz3o2vFJCX230h7' by interface: 'token-identity-provider'

 CONJ00098I Successfully found JWT identity 'host/apps/00uksxz3o2vFJCX230h7'

```

File: app/domain/authentication/authn_jwt/consts.rb (hunk 20 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/consts.rb:16]: No, we have consts files only for authenticators

File: app/domain/errors.rb [hunk 21]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
         code: "CONJ00098E"
       )
 
+      MissingIdentity = ::Util::TrackableErrorClass.new(
+        msg: "JWT identity is empty or was not found.",
```
Comment (shulifink) [inline app/domain/errors.rb:478]: fine

File: app/domain/errors.rb [hunk 22]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
...
+        code: "CONJ00101E"
+      )
+
+      MissingIdentityPrefix = ::Util::TrackableErrorClass.new(
+        msg: "JWT identity prefix is empty or was not found.",
```
Comment (shulifink) [inline app/domain/errors.rb:483]: fine

File: app/domain/logs.rb [hunk 23]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
         code: "CONJ00109D"
       )
 
+      FetchingIdentityPath = ::Util::TrackableLogMessageClass.new(
+        msg: "Fetching identity path...",
```
Comment (shulifink) [inline app/domain/logs.rb:559]: fine

File: app/domain/logs.rb (hunk 16 above)
Comment (shulifink) [reply app/domain/logs.rb:574]: fine

File: app/domain/logs.rb [hunk 24]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00110D"
+      )
+
+      FetchedIdentityPath = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully fetched JWT identity path '{0-identity-path}'",
```
Comment (shulifink) [inline app/domain/logs.rb:564]: fine

File: app/domain/logs.rb [hunk 25]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00111I"
+      )
+
+      FetchingIdentityByInterface = ::Util::TrackableLogMessageClass.new(
+        msg: "Fetching JWT identity by interface: '{0-interface-name}'...",
```
Comment (shulifink) [inline app/domain/logs.rb:569]: fine

File: app/domain/logs.rb (hunk 17 above)
Comment (shulifink) [reply app/domain/logs.rb:579]: fine

File: app/domain/logs.rb [hunk 26]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00113I"
+      )
+
+      AddingIdentityPrefixToIdentity = ::Util::TrackableLogMessageClass.new(
+        msg: "Adding identity prefix '{0-identity-prefix}' to identity '{1-identity}'...",
+        code: "CONJ00114D"
+      )
+
+      AddedIdentityPrefixToIdentity = ::Util::TrackableLogMessageClass.new(
+        msg: "Successfully added JWT identity prefix. Calculated identity is: '{0-identity}'",
```
Comment (shulifink) [inline app/domain/logs.rb:584]: I'm curiuous about "Calculated". Is it really calculated? Maybe "Configured identity is"/"The full identity is:"

File: app/domain/logs.rb [hunk 27]
```diff
@@ -555,6 +555,41 @@ module AuthnJwt
...
+        code: "CONJ00115D"
+      )
+
+      IdentityPathNotConfigured = ::Util::TrackableLogMessageClass.new(
+        msg: "JWT identity path '{0-resource-name}' not configured. JWT identity will be taken from root policy",
```
Comment (shulifink) [inline app/domain/logs.rb:589]: fine

**Review by shulifink — CHANGES_REQUESTED** [review]
All great. Just the one with "Calculated"... see comment

File: app/domain/logs.rb (hunk 26 above)
Comment (nessiLahav) [reply app/domain/logs.rb:584]: We perform 2 manipulations on the identity:
1. Take the identity string from the token
2. add the prefix from configuration
3. add "host" in order to block human users 

So if in the future we will have more steps the term "full" can be confusing

**Review by tzheleznyak — DISMISSED** [review]
LGTM

**Review by tzheleznyak — APPROVED** [review]

**Pull request merged**